### PR TITLE
iam_role: allow retention of existing policies

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
I rebased and updated #24080 with the requested changes.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/iam_role.py

##### ANSIBLE VERSION
```
2.4.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/iam_role.py [hunk 1]
```diff
@@ -197,6 +203,15 @@ def convert_friendly_names_to_arns(connection, module, policy_names):
         module.fail_json(msg="Couldn't find policy: " + str(e))
 
 
+def remove_policies(connection, module, policies_to_remove, params):
+    for policy in policies_to_remove:
+        try:
+            connection.detach_role_policy(RoleName=params['RoleName'], PolicyArn=policy)
+        except ClientError as e:
+             module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
```
Comment (billwanjohi) [inline lib/ansible/modules/cloud/amazon/iam_role.py]: looks like this is what ansibot doesn't like; one too many spaces

**Review by billwanjohi — COMMENTED** [review]
Other than the pep8 indentation problem, this looks good to me and in the spirit of my original submission. Hopefully someone will merge your PR before another rando wanders by and asks us to fix some more unrelated problems...

File: lib/ansible/modules/cloud/amazon/iam_role.py (hunk 1 above)
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/iam_role.py]: Thanks for looking at this! Will fix shortly.

File: lib/ansible/modules/cloud/amazon/iam_role.py [hunk 2]
```diff
@@ -47,6 +47,12 @@
         To embed an inline policy, use M(iam_policy). To remove existing policies, use an empty list item.
     required: true
     aliases: ['managed_policies']
+  remove_unlisted_policies:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/iam_role.py]: Can you name this "purge_policies" or "purge_unlisted_policies" instead to match the SG tags and ec2 modules?

**Review by ryansb — DISMISSED** [review]

**Review by billwanjohi — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/iam_role.py [hunk 3]
```diff
@@ -355,7 +363,8 @@ def main():
             path=dict(default="/", type='str'),
             assume_role_policy_document=dict(type='json'),
             managed_policy=dict(type='list', aliases=['managed_policies']),
-            state=dict(choices=['present', 'absent'], required=True)
+            state=dict(choices=['present', 'absent'], required=True),
+            purge_policies=dict(type='bool', default=True),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/iam_role.py]: Should we be defaulting to True here? I know it's current behaviour but I'd think we might want to protect people from unintentional self-harm:

https://twitter.com/stigglor/status/888221882360582145

**Comment by willthames** [comment]
I see that @wimnat disagrees with my thoughts on default for purge_policies in #24080. We should probably decide on this.

**Comment by s-hertel** [comment]
@willthames I discussed this with @ryansb and it was decided this should default to true so backwards compatibility isn't broken.

**Review by ryansb — APPROVED** [review]

**Pull request merged**